File: Packages/StrandDesign/Sources/StrandDesign/WhoopChrome.swift
import SwiftUI

// MARK: - WhoopChrome
//
// The shared WHOOP page furniture that sits around the RingDial: the full-bleed slate
// canvas, the metric list rows, the "Today vs. prior 30 days" footer pill, the insight
// callout card, the two-up monitor tiles, and the tracked-caps label helper. All of it
// reads from StrandPalette / StrandFont only (no raw hexes), so a palette re-theme flows
// through automatically. Matches reference frames 01, 03 and 04.

// MARK: - TrackedCapsLabel
//
// Small ALL-CAPS Montserrat semibold with wide tracking: the WHOOP section/row label.
// Used by MetricRow, MonitorTile and any screen that needs a house caps label.

/// A tracked ALL-CAPS Montserrat-semibold label (the WHOOP row/section label).
public struct TrackedCapsLabel: View {
    public var text: String
    public var size: CGFloat
    public var color: Color

    public init(_ text: String, size: CGFloat = 11, color: Color = StrandPalette.textSecondary) {
        self.text = text
        self.size = size
        self.color = color
    }

    public var body: some View {
        Text(text)
            .font(StrandFont.text(size, weight: .semibold))
            .tracking(size * 0.12)
            .textCase(.uppercase)
            .foregroundStyle(color)
    }
}

// MARK: - CanvasBackground
//
// The full-bleed WHOOP slate gradient (#283339 top to #101518 bottom). Replaces BOTH the
// flat navy background and the old animated sky sunset header everywhere. Non-interactive
// and hidden from accessibility (pure decoration). Drop it in a `.background { }` or a ZStack.

/// The full-bleed WHOOP slate canvas gradient. Full-bleed by default (ignores safe area).
public struct CanvasBackground: View {
    public init() {}

    public var body: some View {
        StrandPalette.canvasGradient
            .ignoresSafeArea()
            .allowsHitTesting(false)
            .accessibilityHidden(true)
    }
}

// MARK: - MetricRow
//
// One row of a WHOOP metric list: a leading line icon, a tracked caps label, a right-aligned
// D-DIN value, and an up / down / steady trend mark. A 1px hairline separator sits under the
// row (suppressed on the last row via `showsSeparator: false`). Frames 03 / 04.

/// The direction of a metric's trend vs. its baseline.
public enum MetricTrend {
    case up      // green up triangle
    case down    // amber down triangle
    case steady  // neutral grey dot
}

/// A single WHOOP metric list row: icon + caps label + value + trend mark, over a hairline.
public struct MetricRow: View {
    public var systemImage: String?
    public var label: String
    public var value: String
    public var trend: MetricTrend?
    public var showsSeparator: Bool

    public init(
        label: String,
        value: String,
        systemImage: String? = nil,
        trend: MetricTrend? = nil,
        showsSeparator: Bool = true
    ) {
        self.label = label
        self.value = value
        self.systemImage = systemImage
        self.trend = trend
        self.showsSeparator = showsSeparator
    }

    public var body: some View {
        VStack(spacing: 0) {
            HStack(spacing: 12) {
                if let systemImage {
                    Image(systemName: systemImage)
                        .font(.system(size: 16, weight: .medium))
                        .foregroundStyle(StrandPalette.textSecondary)
                        .frame(width: 22, alignment: .center)
                }
                TrackedCapsLabel(label, size: 12, color: StrandPalette.textPrimary)
                Spacer(minLength: 12)
                Text(value)
                    .font(StrandFont.number(17, weight: .bold))
                    .foregroundStyle(StrandPalette.textPrimary)
                if let trend {
                    trendMark(trend)
                }
            }
            .padding(.vertical, 14)
            if showsSeparator {
                Rectangle()
                    .fill(StrandPalette.hairline)
                    .frame(height: 1)
            }
        }
        .accessibilityElement(children: .combine)
    }

    @ViewBuilder
    private func trendMark(_ trend: MetricTrend) -> some View {
        switch trend {
        case .up:
            Image(systemName: "arrowtriangle.up.fill")
                .font(.system(size: 10))
                .foregroundStyle(StrandPalette.statusPositive)
        case .down:
            Image(systemName: "arrowtriangle.down.fill")
                .font(.system(size: 10))
                .foregroundStyle(StrandPalette.statusWarning)
        case .steady:
            Circle()
                .fill(StrandPalette.textTertiary)
                .frame(width: 5, height: 5)
        }
    }
}

// MARK: - TrendFooterPill
//
// The dark rounded pill under a metric list: two trend triangles (green up, amber down)
// then a bold leading word and a secondary trailing phrase, e.g. "Today vs. prior 30 days".
// Frames 03 / 04.

/// The "Today vs. prior 30 days" footer pill: trend legend + bold/secondary caption.
public struct TrendFooterPill: View {
    public var leading: String
    public var trailing: String

    public init(leading: String = "Today", trailing: String = "vs. prior 30 days") {
        self.leading = leading
        self.trailing = trailing
    }

    public var body: some View {
        HStack(spacing: 8) {
            Image(systemName: "arrowtriangle.up.fill")
                .font(.system(size: 9))
                .foregroundStyle(StrandPalette.statusPositive)
            Image(systemName: "arrowtriangle.down.fill")
                .font(.system(size: 9))
                .foregroundStyle(StrandPalette.statusWarning)
            Text(leading)
                .font(StrandFont.text(13, weight: .semibold))
                .foregroundStyle(StrandPalette.textPrimary)
            Text(trailing)
                .font(StrandFont.text(13, weight: .regular))
                .foregroundStyle(StrandPalette.textSecondary)
            Spacer(minLength: 0)
        }
        .padding(.horizontal, 14)
        .padding(.vertical, 10)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            RoundedRectangle(cornerRadius: 10, style: .continuous)
                .fill(StrandPalette.surfaceInset)
        )
        .accessibilityElement(children: .combine)
    }
}

// MARK: - InsightCallout
//
// A dark card with a thin accent border, a line of prose, and a tracked-caps link with a
// trailing arrow (the WHOOP "BREAK DOWN MY RECOVERY ->" affordance). The accent defaults to
// the CTA teal; recovery/strain screens can pass their domain tint. Frame 03 / 04.

/// A dark insight card: thin accent border, prose, and a caps link with a trailing arrow.
public struct InsightCallout: View {
    public var message: String
    public var linkTitle: String
    public var accent: Color
    public var action: (() -> Void)?

    public init(
        message: String,
        linkTitle: String,
        accent: Color = StrandPalette.accent,
        action: (() -> Void)? = nil
    ) {
        self.message = message
        self.linkTitle = linkTitle
        self.accent = accent
        self.action = action
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text(message)
                .font(StrandFont.body)
                .foregroundStyle(StrandPalette.textPrimary)
                .fixedSize(horizontal: false, vertical: true)
            link
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            RoundedRectangle(cornerRadius: NoopMetrics.cardRadius, style: .continuous)
                .fill(StrandPalette.surfaceRaised)
        )
        .overlay(
            RoundedRectangle(cornerRadius: NoopMetrics.cardRadius, style: .continuous)
                .strokeBorder(accent.opacity(0.55), lineWidth: 1)
        )
    }

    private var link: some View {
        let content = HStack(spacing: 6) {
            TrackedCapsLabel(linkTitle, size: 12, color: accent)
            Image(systemName: "arrow.right")
                .font(.system(size: 12, weight: .semibold))
                .foregroundStyle(accent)
        }
        return Group {
            if let action {
                Button(action: action) { content }
                    .buttonStyle(.plain)
            } else {
                content
            }
        }
    }
}

// MARK: - MonitorTile
//
// A two-up caps-header tile (HEALTH MONITOR / STRESS MONITOR): a tracked caps title with an
// optional trailing chevron, then a content slot. Place two side by side in an HStack for the
// WHOOP two-up row. Frame 01.

/// A WHOOP monitor tile: caps header (optional chevron) over a content slot.
public struct MonitorTile<Content: View>: View {
    public var title: String
    public var showsChevron: Bool
    @ViewBuilder public var content: () -> Content

    public init(
        title: String,
        showsChevron: Bool = true,
        @ViewBuilder content: @escaping () -> Content
    ) {
        self.title = title
        self.showsChevron = showsChevron
        self.content = content
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            HStack(spacing: 4) {
                TrackedCapsLabel(title, size: 11, color: StrandPalette.textSecondary)
                Spacer(minLength: 4)
                if showsChevron {
                    Image(systemName: "chevron.right")
                        .font(.system(size: 11, weight: .semibold))
                        .foregroundStyle(StrandPalette.textTertiary)
                }
            }
            content()
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(
            RoundedRectangle(cornerRadius: NoopMetrics.cardRadius, style: .continuous)
                .fill(StrandPalette.surfaceRaised)
        )
        .overlay(
            RoundedRectangle(cornerRadius: NoopMetrics.cardRadius, style: .continuous)
                .strokeBorder(StrandPalette.hairline, lineWidth: 1)
        )
    }
}

#if DEBUG
#Preview("WhoopChrome") {
    ScrollView {
        VStack(spacing: 20) {
            HStack(spacing: 12) {
                MonitorTile(title: "Health Monitor") {
                    HStack(spacing: 8) {
                        Image(systemName: "checkmark.circle.fill")
                            .foregroundStyle(StrandPalette.statusPositive)
                        VStack(alignment: .leading, spacing: 2) {
                            TrackedCapsLabel("Within Range", size: 10, color: StrandPalette.statusPositive)
                            Text("5/5 Metrics")
                                .font(StrandFont.subhead)
                                .foregroundStyle(StrandPalette.textSecondary)
                        }
                    }
                }
                MonitorTile(title: "Stress Monitor") {
                    HStack(spacing: 8) {
                        Text("1.5")
                            .font(StrandFont.number(15, weight: .bold))
                            .foregroundStyle(StrandPalette.stressColor)
                        VStack(alignment: .leading, spacing: 2) {
                            TrackedCapsLabel("Medium", size: 10, color: StrandPalette.textPrimary)
                            Text("4:31pm")
                                .font(StrandFont.subhead)
                                .foregroundStyle(StrandPalette.textSecondary)
                        }
                    }
                }
            }

            VStack(spacing: 0) {
                MetricRow(label: "Heart Rate Variability", value: "124",
                          systemImage: "waveform.path.ecg", trend: .up)
                MetricRow(label: "Resting Heart Rate", value: "49",
                          systemImage: "heart", trend: .up)
                MetricRow(label: "Respiratory Rate", value: "14.5",
                          systemImage: "lungs", trend: .steady, showsSeparator: false)
            }
            .padding(.horizontal, 16)
            .background(
                RoundedRectangle(cornerRadius: NoopMetrics.cardRadius, style: .continuous)
                    .fill(StrandPalette.surfaceRaised)
            )

            TrendFooterPill()

            InsightCallout(
                message: "Your HRV is elevated while your RHR, Respiratory Rate, and Sleep Performance are all typical, resulting in a higher Recovery today.",
                linkTitle: "Break down my recovery"
            )
        }
        .padding(20)
    }
    .frame(width: 420, height: 720)
    .background(CanvasBackground())
    .preferredColorScheme(.dark)
}
#endif

File: android/app/src/main/java/com/noop/ui/components/WhoopChrome.kt
package com.noop.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.noop.ui.Metrics
import com.noop.ui.NoopType
import com.noop.ui.Palette

// MARK: - WhoopChrome (the shared WHOOP layout furniture, Compose parity with StrandDesign/WhoopChrome.swift)
//
// The WHOOP "chrome" that every restyled screen composes on top of the RingDial:
//   1. CanvasBackground  the slate vertical-gradient page background (replaces LiquidSky everywhere).
//   2. TrackedCapsLabel  the small uppercase, wide-tracking Montserrat SemiBold label used app-wide.
//   3. MetricRow         icon + tracked caps label + right value + up/down trend triangle, hairline rule.
//   4. TrendFooterPill   the "Today vs. prior 30 days" comparison pill under a metric list.
//   5. InsightCallout    a dark card with a thin teal/blue border and a caps link with an arrow.
//   6. MonitorTile       the two-up HEALTH MONITOR / STRESS MONITOR tile pattern.
//
// Colour is ONLY palette tokens (no raw hexes). Mirrors the Swift contract and the reference screenshots.

// MARK: - CanvasBackground

/**
 * The WHOOP slate page background: a vertical gradient from lighter slate (#283339) at the top to
 * near-black (#101518) at the bottom, built from [Palette.canvasGradient]. Fills its box and lays
 * [content] over the gradient. Replaces the old animated LiquidSky header AND the flat navy canvas on
 * every screen. Mirrors the Swift `CanvasBackground` view.
 */
@Composable
fun CanvasBackground(
    modifier: Modifier = Modifier,
    content: @Composable BoxScope.() -> Unit = {},
) {
    Box(
        modifier = modifier
            .fillMaxSize()
            .background(Palette.canvasGradient()),
    ) {
        content()
    }
}

// MARK: - TrackedCapsLabel

/**
 * The WHOOP tracked-caps label: uppercase, wide letter-tracking, Montserrat SemiBold, secondary grey by
 * default. The one building block behind every caps label in the chrome (metric rows, monitor tiles,
 * footer pills). [size] is the point size; tracking scales with it. Mirrors the Swift `TrackedCapsLabel`.
 */
@Composable
fun TrackedCapsLabel(
    text: String,
    modifier: Modifier = Modifier,
    color: Color = Palette.textSecondary,
    size: Float = 11f,
    weight: FontWeight = FontWeight.SemiBold,
) {
    Text(
        text = text.uppercase(),
        style = NoopType.text(size, weight).copy(letterSpacing = (size * 0.14f).sp),
        color = color,
        maxLines = 1,
        overflow = TextOverflow.Ellipsis,
        modifier = modifier,
    )
}

// MARK: - MetricRow

/** The direction of a [MetricRow]'s trend triangle. */
enum class MetricTrend {
    /** Rising: a green up triangle. */
    Up,

    /** Falling: a red down triangle. */
    Down,

    /** Flat / neutral: a muted dash. */
    Flat,

    /** No trend shown at all. */
    None,
}

/**
 * A single WHOOP metric row: an optional leading [icon], the tracked-caps [label], then the right-aligned
 * [value] (D-DIN) and an optional up/down [trend] triangle. A hairline rule sits under the row unless
 * [showDivider] is false (the last row in a list drops it). Mirrors the Swift `MetricRow`.
 */
@Composable
fun MetricRow(
    label: String,
    value: String,
    modifier: Modifier = Modifier,
    icon: ImageVector? = null,
    trend: MetricTrend = MetricTrend.None,
    valueColor: Color = Palette.textPrimary,
    showDivider: Boolean = true,
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .then(
                if (showDivider) {
                    Modifier.drawBehind {
                        val y = size.height
                        drawLine(
                            color = Palette.hairline,
                            start = Offset(0f, y),
                            end = Offset(size.width, y),
                            strokeWidth = Metrics.divider.toPx(),
                        )
                    }
                } else {
                    Modifier
                },
            )
            .padding(vertical = 14.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        if (icon != null) {
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = Palette.textSecondary,
                modifier = Modifier.size(Metrics.iconSmall),
            )
        }
        TrackedCapsLabel(text = label, modifier = Modifier.weight(1f))
        Text(
            text = value,
            style = NoopType.number(17f, FontWeight.SemiBold),
            color = valueColor,
            maxLines = 1,
        )
        val triangle: Pair<String, Color>? = when (trend) {
            MetricTrend.Up -> "▲" to Palette.statusPositive
            MetricTrend.Down -> "▼" to Palette.statusCritical
            MetricTrend.Flat -> "-" to Palette.textTertiary
            MetricTrend.None -> null
        }
        if (triangle != null) {
            Text(
                text = triangle.first,
                style = NoopType.captionNumber.copy(fontSize = 9.sp, fontWeight = FontWeight.Bold),
                color = triangle.second,
            )
        }
    }
}

// MARK: - TrendFooterPill

/**
 * The comparison pill that sits under a metric list, e.g. "Today vs. prior 30 days". A rounded, inset pill
 * with a tracked-caps label centred inside a hairline border. Mirrors the Swift `TrendFooterPill`.
 */
@Composable
fun TrendFooterPill(
    text: String = "Today vs. prior 30 days",
    modifier: Modifier = Modifier,
) {
    val shape = RoundedCornerShape(Metrics.cornerPill)
    Box(
        modifier = modifier
            .clip(shape)
            .background(Palette.surfaceInset)
            .border(1.dp, Palette.hairline, shape)
            .padding(horizontal = 14.dp, vertical = 7.dp),
        contentAlignment = Alignment.Center,
    ) {
        TrackedCapsLabel(text = text, color = Palette.textTertiary, size = 10f)
    }
}

// MARK: - InsightCallout

/**
 * A WHOOP insight callout: a dark card with a thin [tint] (teal / blue) border carrying a body [text] and
 * an optional caps [linkLabel] with a trailing arrow. Tapping fires [onClick] (only wired when a link is
 * present). Mirrors the Swift `InsightCallout`.
 */
@Composable
fun InsightCallout(
    text: String,
    modifier: Modifier = Modifier,
    linkLabel: String? = null,
    tint: Color = Palette.accent,
    onClick: (() -> Unit)? = null,
) {
    val shape = RoundedCornerShape(Metrics.cardRadius)
    Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(shape)
            .background(Palette.surfaceRaised)
            .border(1.dp, tint.copy(alpha = 0.55f), shape)
            .then(if (onClick != null) Modifier.clickable(onClick = onClick) else Modifier)
            .padding(18.dp),
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Text(
            text = text,
            style = NoopType.subhead,
            color = Palette.textSecondary,
        )
        if (linkLabel != null) {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(6.dp),
            ) {
                TrackedCapsLabel(text = linkLabel, color = tint, size = 11f, weight = FontWeight.Bold)
                Text(
                    text = "→",
                    style = NoopType.caption.copy(fontWeight = FontWeight.Bold),
                    color = tint,
                )
            }
        }
    }
}

// MARK: - MonitorTile

/**
 * A WHOOP monitor tile: the tracked-caps [title] (e.g. "HEALTH MONITOR"), a big D-DIN [value] and an
 * optional [caption]. An optional [content] slot sits below the value for a sparkline or status row. A
 * [tint] colours the title toward its domain world. Compose two side by side in a weighted Row for the
 * two-up HEALTH MONITOR / STRESS MONITOR pattern. Mirrors the Swift `MonitorTile`.
 */
@Composable
fun MonitorTile(
    title: String,
    value: String,
    modifier: Modifier = Modifier,
    caption: String? = null,
    tint: Color = Palette.accent,
    valueColor: Color = Palette.textPrimary,
    content: (@Composable ColumnScope.() -> Unit)? = null,
) {
    val shape = RoundedCornerShape(Metrics.cardRadius)
    Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(shape)
            .background(Palette.surfaceRaised)
            .border(1.dp, Palette.hairline, shape)
            .padding(Metrics.cardPadding),
        verticalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        TrackedCapsLabel(text = title, color = tint, size = 11f)
        Text(
            text = value,
            style = NoopType.number(26f, FontWeight.Bold),
            color = valueColor,
            maxLines = 1,
        )
        if (caption != null) {
            Text(
                text = caption,
                style = NoopType.footnote,
                color = Palette.textTertiary,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
        }
        if (content != null) {
            content()
        }
    }
}
